### PR TITLE
WIP: Dockerfile and development docker-compose setup.

### PR DESCRIPTION
**Pull request by bemosior**
This is a work-in-progress PR to add Docker support to VuFind. I based the initial draft of these changes on the latest 3.1 tag, so I could start from a known-working VuFind state (that's also why the 3.1.3 version bump commit is caught up in this PR; let me know if that's an issue). For a completely working example context to play with, see [unicalabs/vufind:release-3.1](https://github.com/unicalabs/vufind/tree/release-3.1).

### Additions

* `Dockerfile` enables the building of a Docker image with support for PHP 7.1 and Apache.
* `docker-vufind-entrypoint` enables custom behavior within the VuFind container at runtime, often based on environment variables (such as `VUFIND_MYSQL_URL` or `VUFIND_SOLR_URL`).
* `docker-compose.yml` enables the creation of a full working stack for development purposes, including VuFind (from the above image), Solr (from their official image), and MySQL (from their official image). The same principles can be applied for a production stack.
* `solr/docker/dev_link_vendor.sh` is a symlink hack to map the file structure the biblio core config expects to the one provided by the official Solr container image.

### Modifications
* `build.xml` now checks if the `VUFIND_INSTALL_SOLR` environment variable is set before installing Solr. If the value is `no`, Solr will not be downloaded and installed. The Docker image doesn't need a Solr server installed within (it has a separate container from the official Solr project for that), so this saves space/time.  

### Development Use
With [Docker](https://www.docker.com/community-edition#/download) and [Docker Compose](https://docs.docker.com/compose/install/) installed, one can run `docker-compose up` and have a fully-working development environment, including a provisioned MySQL database (available from your local machine at `mysql://root@localhost/vufind` and within the VuFind container at `mysql://root@mysql/vufind`) and a Solr instance (available from your local machine at http://localhost:8983/solr and within the VuFind container at http://solr:8983/solr).

**Warning:** By design, data in the MySQL database and Solr instance currently will disappear when their respective containers are halted.

### Production Use
Production use looks a bit different. As I work through the implementation, I'd like to create some documentation and public examples.

### To-Do
A few things I need to work on:

- [x]  Resolve Zend error, likely from post-`3.1.2` work-in-process in `master` (I could use help with this one):
```
Fatal error: Uncaught TypeError: Argument 1 passed to Zend\ServiceManager\AbstractPluginManager::__construct() must implement interface Zend\ServiceManager\ConfigInterface or be null, instance of Zend\ServiceManager\ServiceManager given, called in /usr/local/vufind/module/VuFind/src/VuFind/Service/Factory.php on line 143 and defined in /usr/local/vufind/vendor/zendframework/zendframework/library/Zend/ServiceManager/AbstractPluginManager.php:62

Stack trace: 
#0 /usr/local/vufind/module/VuFind/src/VuFind/Service/Factory.php(143): Zend\ServiceManager\AbstractPluginManager->__construct(Object(Zend\ServiceManager\ServiceManager), Array) 
#1 [internal function]: VuFind\Service\Factory::getConfig(Object(Zend\ServiceManager\ServiceManager), 'vufindconfig', 'VuFind\\Config') 
#2 /usr/local/vufind/vendor/zendframework/zendframework/library/Zend/ServiceManager/ServiceManager.php(939): call_user_func('VuFind\\Service\\...', Object(Zend\ServiceManager\ServiceManager), 'vufindconfig', 'VuFind\\Config') 
#3 /usr/local/vufind/vendor/zendfram in /usr/local/vufind/vendor/zendframework/zendframework/library/Zend/ServiceManager/AbstractPluginManager.php on line 62
```
- [x]  Resolve Solr error, likely just a moved library from post-`3.1.2` work-in-process in `master`:
```
org.apache.solr.common.SolrException: Could not load conf for core biblio: Can't load schema /opt/solr/server/solr/mycores/biblio/conf/schema.xml: Plugin init failure for [schema.xml] fieldType "text": Plugin init failure for [schema.xml] analyzer/tokenizer: Error loading class 'solr.ICUTokenizerFactory'
```
- [x]  Improve `docker-vufind-entrypoint` to hot-overwrite the local `config.ini` with MySQL/Solr connection information from environment variables. The default state needs to work for dev, and prod use should necessitate setting these values.
- [ ]  Create example project: https://github.com/unicalabs/vufind-docker-example
- [x]  Revise according to [failed Codacy checks](https://www.codacy.com/app/crhallberg/vufind/pullRequest?prid=639234).
- [x]  Make VuFind available at the site root instead of at `/vufind`.
- [x]  Adjust the `phing installsolr` job to do nothing if it's running in a dockerized environment. This should improve image build and dev first-run performance significantly.

### CODE REVIEW

**Comment by lahmann** [comment]
Great to see official docker support in VuFind! Maybe you want to take a look at our VuFind docker image (mainly for development, not production): https://github.com/finc/docker-vufind2
Best,
André

**Comment by useltmann** [comment]
Nice work there! Im very happy to see where you are going with respect to production use. Splitting the services, using the official images and so on. However i would not try to cover development usage as well. In my opinion this is *a)* overwhelming the developer by taking care of multiple images/containers/networks/... and *b)* the container-configuration scripts become too complex to cover both scenarios. For example the current setup is not respecting the developers uid/gid, which makes it difficult to remove temporary files created by the services. Also in order to run the container one has to provide the vufind-code in advance. This is problematic in fully-automated production deployment processes.
I would recommend (a phing task, better github-hook to docker-hub) to pack the vufind-code and all dependencies inside an image.
Next problem: the vufind-apache2-config is set up for a http-path `/vufind` which is suboptimal in my opinion. All of our catalogue deployments are in subdomain-context so something like this should be configurable. 
Let me see whether i can provide you with a PR.

**Comment by demiankatz** [comment]
@bemosior, I had neglected to merge the latest version bumps from release-3.1 into master. I just took care of that, and it makes this PR more readable by eliminating the irrelevant noise.

I'm not sure why you are seeing that Zend error. It looks like the sort of thing that can happen when mixing incompatible versions of Zend Framework components, but I"m not sure why that would happen on a fresh install since as far as I know, composer.json should be very specific in every branch to avoid these kinds of problems.

The Solr error is also a bit odd, but it looks like Solr is trying to find the schema in a weird place.

I would actually recommend trying to merge the latest master into this to see if it makes any difference; with the 4.0 release coming in July, I don't expect a lot of major churn in master over the next couple of months, and it's probably best to target the more up-to-date, forward-looking code.

If you're still stuck after that, let me know if anything has changed and I'll give it some more thought!

**Comment by demiankatz** [comment]
(Also, apologies for messing around with this branch -- not sure if you noticed, but I pushed something up and immediately retracted it when I realized it was a mistake).

**Comment by bemosior** [comment]
I appreciate the pointer, @lahmann! Thanks to @useltmann and @ckaz, that repository was one of my reference points for this process!

Thanks, @useltmann, for the useful suggestions by email and the helpful feedback on this PR!

### Development use case
I believe providing support for development is important, and I would like to do whatever it takes to make it reasonable. To avoid overwhelming the developer, I highly recommend using Docker Compose, which makes use of the `docker-compose.yml` file. Or am I misunderstanding? I'd like to make sure I grasp the problem fully.

### Development/production conflict
As far as the configuration scripts being too rigid for both development and production use, I agree. I would eventually like to commit a `docker-compose-production.yml` that offers a reasonable baseline for production use, but first there's an important paradigm shift to be explored regarding VuFind development, customization, and production deployment practices when it's all Dockerized. For example, one could have a repository containing only files from the VuFind project that have been customized. With Docker, they could just be overlayed on top of an official Docker image instead of forking the entire project.

I'm not sure I'm explaining it the best, which is why I would like to build out a few examples at https://github.com/unicalabs/vufind-docker-example. I am using a container management service ([Rancher](http://rancher.com/)), so I will certainly have a better idea of what needs to happen for practical production use in the near future.

### Builds
Yes! Exactly, you got it. As a test, I've configured automated builds of the Unica VuFind project under the `unica/vufind` namespace. https://hub.docker.com/r/unica/vufind/

Once this is eventually merged in, @demiankatz could create a VuFind org on Docker Hub, link to his GitHub account, and create an automated build for the VuFind project. It automatically watches for pushes to any branch on GitHub and tags successful builds ([example](https://hub.docker.com/r/unica/vufind/tags/)).

### Http path
I knew I forgot something! I meant to include this change in the first place but forgot. I've added it to the to-do list above. Thanks!

**Comment by bemosior** [comment]
Thanks, @demiankatz I'll merge in the latest master shortly.

I'll take a look at the Zend issue more closely. Perhaps there's something obvious I'm missing.

The Solr error is one I've fixed in the past. It should just be a matter of telling the container to put the right library in the right place. I'll do that as soon as I can.

And no worries about the branch. You can mess with my branches anytime! :)

**Comment by demiankatz** [comment]
@bemosior, sounds good. Regarding the Zend issue, it's the sort of thing that I sometimes run into as a result of caching or Composer problems -- but those seem unlikely in the scenario of building a brand-new Docker container. Anyway, let me know if I can be of more assistance!

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,21 @@
 FROM php:7.1-apache
+MAINTAINER "demian.katz@villanova.edu"
```
Comment (bemosior) [inline Dockerfile]: @demiankatz I'm happy to put my contact info here if you don't want to be on the hook for the Docker stuff quite yet. :)

**Comment by bemosior** [comment]
The comment about the Zend error being the result of a composer issue was totally correct.

Right now we bake the `vendor` folder into the docker image, but to use Docker for development we want also want to override the VuFind folder. So then our local workspace `vendor` overrides the baked-in `vendor`, and VuFind gets upset because it might not be the right set of dependencies.

The solution was just to conditionally run `composer install` within the container on spin-up, which should make our local workspace `vendor` folder reflect what the container expects. It takes some time on first-run, but future runs are quick because there's nothing to do.

I'd like to significantly improve image build and first-run dev performance by adding a condition to the `phing installsolr` job where it does nothing if it's in a dockerized environment. I think there should be an easy way to do that within `build.xml`.

**Comment by useltmann** [comment]
Sorry, @bemosior, but unfortunately i will not be able to provide you with a decent PR in near future. Im almost off for vacation and not back until june.

The **uid/gid** issue is not that simple. AFAIR apache is not allowing to run as user with a uid/gid higher or equal than 1000(?) which would be required if you run apache2 with *php_mod* support and create files so that the user has sufficient permissions. 
Instead in my developer-image i choosed to run apache2 with *php-fpm* support so i could adjust the user of the fpm-process. I suggest you have a look at https://github.com/useltmann/docker-dev-dotdeb/blob/master/assets/setup. The `01-appuser-uid` especially detects the developers uid/gid from the bind folder's owner and adjusts the app-users uid and gid.
There is some more magic going on but see for yourself. In generally i can say that all script in folder `build` are run on image-build. They do *not* apply on image-builds of inherited images. All scripts in folder `setup` are executed on each container-start, no matter if its the first or the hundredst.

I hope that halps a bit.
Cheers

**Comment by bemosior** [comment]
@useltmann Sorry I let this wait for so long. I do not yet understand the UID/GID issue you describe (I'm probably just being dense). I'll be able to get it more quickly if I can see the issue myself. Could you help me figure out how to cause the issue with the code in this PR?

**Comment by useltmann** [comment]
@bemosior Coming back to you almost after a year. Sorry about that but i was busy with other things.

Regarding the UID/GID issue, the problem is, that you just cannot delete the vendor-files or other files created by vufind as the developer you are. I find it quite annoying to become root in order to clean my working area.

I come to the conclusion that vufinds service-structure is far to complex to have it all in only one image. 
Therefore i am working on a set of images to use for vufind in production- and development-environments focused on orchestration-tools like docker-compose, rancher and even kubernetes.
This not only makes it flexible to use and switch components to the ones you favor. Also it helps to remove complexity and adds the ability to scale horizontally.

Also I created a tool to autoconfigure vufind (create database, create config-inheritance, customise specific variables) in our deploy-context. This tool is included inside the image and enables us to switch vufind-branches (in context of libraries) and start the entire setup with no interaction needed. you can find the sources here: [vufind-php], [vufind-httpd] [vufind-nginx]

The images are fairly simple, maybe the vufind-php-Image might be not, but it is also doing the most work. The tool which is responsible for creating the configs and database can be found here: [@ubleipzig/autoconfig] ([docs]).

We use *compose* to start the docker-containers. Here is an example one of our `docker-compose.yml`:
```yaml
version: '2'
services:
  php:
    image: ubleipzig/vufind-php:5.6-8-debug
    volumes:
      - ./:/usr/local/vufind
    environment:
      VUFIND_LOCAL_DIR: /usr/local/vufind/local/dev
      VUFIND_CACHE_DIR: /usr/local/vufind/data/cache
      VUFIND_LOCAL_MODULES: finc
      VUFIND_ENV: development

  httpd:
    image: ubleipzig/vufind-httpd:2.4-2
    ports:
      - 80:80
    volumes:
      - ./:/usr/local/vufind:ro
      - ./data/cache:/var/cache/vufind:ro
    depends_on:
      - php
    command: httpd -D FOREGROUND -c "ProxyTimeout 3600"

  db:
    image: mariadb:10.2
    ports:
      - 127.0.0.1:3306:3306
    volumes:
      - mariadb-data:/var/lib/mysql:z
    environment:
      MYSQL_ROOT_PASSWORD: adminpw
    command:
      - --character-set-server=utf8mb4
      - --collation-server=utf8mb4_unicode_ci

  smtp:
    image: useltmann/mailcollect:8-1
    ports:
      - '143:143'
      - '25:25'
    volumes:
      - mail-data:/home/dev/Maildir:z

  grunt:
    image: ubleipzig/vufind-php:5.6-8-debug
    volumes:
      - ./:/usr/local/vufind:z
      - npm-cache:/home/www-data/.npm
    environment:
      NODE_ENV: development
    command: npm install && node_modules/.bin/grunt watch

  composer:
    image: ubleipzig/vufind-php:7.2-8-debug
    volumes:
      - ./:/usr/local/vufind:z
      - composer-cache:/home/www-data/.composer
      - ${SSH_AUTH_SOCK}:/ssh-agent
    environment:
      SSH_AUTH_SOCK: /ssh-agent
    command: phing composer -Dcomposer_extra_params=--no-scripts

  autoconfig:
    image: ubleipzig/vufind-php:5.6-8-debug
    volumes:
      - ./:/usr/local/vufind:z
    environment:
      VUFIND_SITE: local
      VUFIND_INSTANCE: dev
      VUFIND_DB_CLIENT: "%"
      VUFIND_DB_ADMIN_PASSWORD: adminpw
      VF_config_ini__Authentication__hash_passwords: "true"
      VF_config_ini__Authentication__encrypt_ils_password: "true"
      VF_config_ini__Database__database: mysql://vufind:vufindpw@db/vufind
      VF_config_ini__Index__url: http://172.18.113.12:8080/solr
      VF_config_ini__Mail__host: smtp
      VF_config_ini__Mail__port: 25
      VF_config_ini__Mail__username: dev
      VF_config_ini__Mail__pasword: dev
    command: sleep 10 && autoconfig vufind deploy
    depends_on:
      - db

volumes:
  composer-cache: {}
  npm-cache: {}
  mail-data: {}
  mariadb-data: {}

networks:
  default:
    driver_opts:
      com.docker.network.bridge.name: vufind-bridge
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 10.2.0.0/16
          gateway: 10.2.0.1
```

I leave you with this information for now. If you have questions, feel free to contact me, I will be happy to explain how it works. 

Next step for us is to enable this setup for continuous delivery via rancher/k8s.

cheers
Ulf

[vufind-php]: https://hub.docker.com/r/ubleipzig/vufind-php
[vufind-httpd]: https://hub.docker.com/r/ubleipzig/vufind-httpd
[vufind-nginx]: https://hub.docker.com/r/ubleipzig/vufind-nginx
[@ubleipzig/autoconfig]: https://www.npmjs.com/package/@ubleipzig/autoconfig
[docs]: https://ubleipzig.github.io/autoconfig/vufind/

**Comment by demiankatz** [comment]
Thanks for sharing this, @useltmann. I'm not sure if @bemosior is still following this, but I'd be interested to hear his thoughts if he is.

What is your advice regarding including Docker configuration with VuFind going forward? Would it be useful, and is it even feasible, to make some or all of your work part of the official VuFind repository so it can be more easily found and reused? Without a detailed knowledge of Docker, I realize that those questions may be misguided. But if there's something I can do to help make your work easier to maintain and share, please let me know. I'm also interested in what the ultimate fate of this pull request should be; is it a lost cause that should be closed, or might we be able to refine this into something that, if shipped with VuFind, would actually be helpful to people?

**Comment by bemosior** [comment]
My $0.02 is that contributed work like this should be integrated as a first-class feature of VuFind if 1) there is significant community interest and adoption, or 2) it enables a strategic imperative of the VuFind project.

Back when I was doing this work, I was pursuing dockerization as a [strategic imperative](https://www.youtube.com/watch?v=WZpE097IR5A) to make the _operation_ of VuFind ridiculously easy... To the extent that most libraries would not need in-house sysadmin expertise dedicated to VuFind, because hosted services could come about that were really cheap. That was a vision I had, and it's still something I view as possible. Dockerization fit into that picture because it enabled those highly efficient hosting solutions (I was thinking about starting one). That was my interest in it.

@useltmann's work enables that kind of future, but it has to be taken the rest of the way. I'm open to having that discussion, as I've done a lot of thinking about it (and still have blueprints).

**Comment by demiankatz** [comment]
Thanks, @bemosior. I think where my lack of experience with Docker lets me down is that I'm not sure strategically where everything should belong. Obviously @useltmann has things out on Docker Hub, which is great for making VuFind visible to people familiar with Docker... but I'm not sure how the versioning works there in terms of making sure that what's on Docker Hub aligns with various releases of VuFind. On the other side, if we include some Docker configurations with VuFind, that clearly ties a working Docker configuration with an appropriate VuFind release... but perhaps visibility or convenience suffers. Maybe there's a hybrid solution to be found. Again, apologies if my ignorance is leading me astray here, but I think you can see the lines along which I am thinking. :-)

**Comment by bemosior** [comment]
To fully support Docker, you would certainly want to integrate Docker image creation (and pushing said images to Docker Hub) into your release process. It's a lot of work to do manually, so you'd want to create an automated process, probably as part of your Continuous Integration process (e.g., Travis CI or something).

But that work is a questionable investment if Dockerized VuFind isn't enabling anything, strategically speaking.

The obvious question is... what does Dockerized VuFind enable? And the short answer is, I think, centralized, inexpensive hosting solutions. And those will likely only be taken advantage of by the minority of institutions who are adopting those hosting paradigms as standard (probably like @useltmann's), as well as those enterprising folks willing to start VuFind hosting services.

My hypothesis is that the above makes VuFind-as-a-Service possible, which is cheap, which in turn could increase VuFind adoption rates, which in turn could increase investment in the project and give the for-profit side of the industry a run for its money. But someone has to be making all that happen. And right now it's not me. If you want to play that kind of long strategy, then you'd want to be paying attention to all of that, all the way.

**Comment by useltmann** [comment]
My advice on integrating docker in VuFind is: Don't. As said i believe that VuFinds ecosystem is far to complex already to put in just a single image. It is also not advisable in concerns of scalablity. My approach is to provide service-images for VuFind (there is acually no VuFind-code inside my images) which can be used together with a working copy of the git repo for developing purposes or a data-image only providing the VuFind-data (which is what we intend to do for deployment).

I recommend to create one (or more) new project(s) besides the VuFind-"code" project similar to what i created. I would be happy if you would like to use them as initial start.

Regarding this PR for me it is just communication. I did not contribute any code. Also i would like to see this conversation somewhere else. The mailinglist?

**Comment by demiankatz** [comment]
@useltmann, thanks for the input -- that is helpful. I think perhaps the Wiki might be the best place to compile some best practices and links to resources related to Docker. I have created a stub page that currently points back to this pull request:

https://vufind.org/wiki/installation:docker

If you would like to share links or instructions there, that would be a great help. If not, of course, I do not wish to impose on your time!

If you would like me to establish a code repository in the vufind-org organization on GitHub where Docker-specific tools could be shared, I would be more than happy to do so. Just let me know, and I can set something up and give you write access to it.

As far as conversation goes, please feel free to post to the mailing lists at any time. I can also reserve time on a future developers call if that would be useful. Ultimately, I think the question is whether anyone else is currently interested in this subject. @bemosior is participating out of his generosity but has no personal stake in it at this point; I am interested in making useful tools available to the VuFind community but do not currently have the need to use Docker myself; I'm not sure if anyone else is actively considering this deployment method. I suppose @lmgonzales might be interested given conversations she had with Ben at the last VuFind Summit. In any case, it might not hurt to reach out to the VuFind lists and poll for interest.

Thanks again for your time and input, and let me know if I can do anything more to assist!

**Comment by demiankatz** [comment]
Because I do not expect this PR to be merged as-is, I am closing it for now. We can always reopen it if the need arises.

**Comment by Eeemil** [comment]
Hello, I'm interested in trying vufind out and I'm quite familiar with Docker so a docker deployment would be neat.
I was directed to this pull request from here: https://vufind.org/wiki/installation:docker 

I just want to get started quickly to see if vufind can parse some MARC records and have a searchable catalog for non-technical users, I don't need a production ready setup with support for all the things. Will https://github.com/finc/docker-vufind suffice?

**Comment by demiankatz** [comment]
@Eeemil, I'm not sure how current any of the Docker-related work is at this stage. The VuFind community as a whole still hasn't really figured out the ideal Docker strategy, so you'll run across a lot of institution-specific work as well as some incomplete proofs of concept... but I'm not sure how well any of them will work with current master code, or what caveats are attached to them. If your goal is to get a VuFind instance up and running quickly so you can play with it, I think trying to use Docker is actually going to slow you down. I'd instead recommend Vagrant (https://vufind.org/wiki/installation:vagrant) for a quick dev instance, since this is known to work with recent VuFind versions and is reasonably well-documented. If you're looking to check out a repo and run a single command to get something up and running, that's your best bet. Installing the latest .deb package on an Ubuntu box is also a pretty painless option.

Of course, I don't mean to dismiss Docker -- I'd still love to have a better way of supporting Docker users; I just don't think that problem has been solved yet (in part because I haven't had time to gain a deep enough understanding of the software). If this is important to you, I'm certainly happy to discuss it further and see if we can get any closer to a solution!

Let me know if I can be of any further assistance!

**Comment by Eeemil** [comment]
Thank you @demiankatz, I'll try the Vagrant install path for now :)

**Pull request closed without merging**